File: python/Greedy/merge_triplets_form_target.py
from typing import List

class Solution:
    def mergeTriplets(self, triplets: List[List[int]], target: List[int]) -> bool:
        """
        Determine if we can form the target triplet by merging some of the given triplets.
        
        Approach:
        - For each triplet, check if any of its values exceed the target values.
        - If a triplet has any value greater than the target, skip it (it can't be used).
        - For valid triplets, check if any of their values match the target values exactly.
        - Keep track of which target positions (0, 1, 2) we can achieve.
        - If we can achieve all three target positions, return True.
        
        Time Complexity: O(N), where N is the number of triplets
        - We iterate through each triplet once
        
        Space Complexity: O(1)
        - The set 'good' stores at most 3 elements
        
        Args:
            triplets: List[List[int]] - List of triplets [a, b, c]
            target: List[int] - Target triplet [x, y, z]
        Returns:
            bool - True if target can be formed by merging triplets, False otherwise
        """
        good = set()

        for t in triplets:
            # Skip triplets that have any value greater than the target
            if t[0] > target[0] or t[1] > target[1] or t[2] > target[2]:
                continue

            # Check if any value in this triplet matches the target exactly
            for i, v in enumerate(t):
                if v == target[i]:
                    good.add(i)
        
        # Return True if we can achieve all three target positions
        return len(good) == 3

# -------------------
# Test Cases
# -------------------
def test_merge_triplets():
    s = Solution()
    # Test case 1: Can form target
    assert s.mergeTriplets([[2,5,3],[1,8,4],[1,7,5]], [2,7,5]) == True, "Failed on [[2,5,3],[1,8,4],[1,7,5]], [2,7,5]"
    # Test case 2: Cannot form target
    assert s.mergeTriplets([[3,4,5],[4,5,6]], [3,2,5]) == False, "Failed on [[3,4,5],[4,5,6]], [3,2,5]"
    # Test case 3: Single triplet matches target
    assert s.mergeTriplets([[1,2,3]], [1,2,3]) == True, "Failed on [[1,2,3]], [1,2,3]"
    # Test case 4: No valid triplets
    assert s.mergeTriplets([[2,2,2],[3,3,3]], [1,1,1]) == False, "Failed on [[2,2,2],[3,3,3]], [1,1,1]"
    # Test case 5: All triplets exceed target
    assert s.mergeTriplets([[4,5,6],[7,8,9]], [1,2,3]) == False, "Failed on [[4,5,6],[7,8,9]], [1,2,3]"
    # Test case 6: Multiple valid triplets
    assert s.mergeTriplets([[1,2,3],[4,2,6],[7,8,3]], [4,8,6]) == False, "Failed on [[1,2,3],[4,2,6],[7,8,3]], [4,8,6]"
    # Test case 7: Empty triplets
    assert s.mergeTriplets([], [1,2,3]) == False, "Failed on [], [1,2,3]"
    print("All test cases passed!")

if __name__ == "__main__":
    test_merge_triplets()

File: C++/Greedy/merge_triplets_form_target.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * Function to check if target triplet can be formed by merging some of the given triplets.
     *
     * Approach:
     * - Iterate over each triplet.
     * - Skip any triplet where a value exceeds the corresponding target value.
     * - For valid triplets, if a value equals the corresponding target value, mark it as achieved.
     * - If we can achieve all three positions (0,1,2), return true.
     *
     * Time Complexity: O(N) — one pass over all triplets.
     * Space Complexity: O(1) — only a few bool flags are used.
     */
    bool mergeTriplets(vector<vector<int>>& triplets, vector<int>& target) {
        // Flags to check if each of the 3 positions has been matched to the target
        bool found0 = false, found1 = false, found2 = false;

        for (const auto& t : triplets) {
            // Skip triplets with any value greater than the corresponding target
            if (t[0] > target[0] || t[1] > target[1] || t[2] > target[2])
                continue;

            // Check which positions of target are achieved by this triplet
            if (t[0] == target[0]) found0 = true;
            if (t[1] == target[1]) found1 = true;
            if (t[2] == target[2]) found2 = true;
        }

        // Return true if all three positions have been achieved
        return found0 && found1 && found2;
    }
};

/*
====================================================
Line-by-line C++ syntax explanation:
====================================================

#include <bits/stdc++.h>
    -> Includes all standard C++ libraries (like <vector>, <iostream>, <string>, etc.)

using namespace std;
    -> Allows using standard library names without writing std:: before them.

class Solution {
public:
    -> Defines a class named Solution with public member functions.

bool mergeTriplets(vector<vector<int>>& triplets, vector<int>& target) {
    -> Defines a method `mergeTriplets` that takes a 2D vector of integers and a vector of integers as input and returns a bool.

bool found0 = false, found1 = false, found2 = false;
    -> Three boolean flags to track if target[0], target[1], and target[2] have been achieved.

for (const auto& t : triplets) {
    -> Range-based for loop. `t` is a reference to each triplet (vector<int>) in `triplets`.

if (t[0] > target[0] || t[1] > target[1] || t[2] > target[2])
    continue;
    -> Skip the current triplet if it has any value greater than the corresponding value in target.

if (t[0] == target[0]) found0 = true;
if (t[1] == target[1]) found1 = true;
if (t[2] == target[2]) found2 = true;
    -> Update flags if this triplet contributes to achieving the target values.

return found0 && found1 && found2;
    -> Return true only if all three flags are true (target can be formed).

====================================================
Time Complexity:
    - O(N), where N is the number of triplets.

Space Complexity:
    - O(1), constant space for the flags.
====================================================
*/

